segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -173,75 +200,68 @@ public int getId() {
-                // which wouldn't have triggered a state.commit on its own.
-                completeNonSerializableOperations(state);
-            }
-        } catch (Throwable ex) {
-            handleIterationException(ex, state, operations);
-        }
-    }
-
-    /**
-     * Processes all the given operations, in order, using the given QueueProcessingState and DataFrameBuilder.
-     *
-     * @param operations       The operations to process.
-     * @param state            The QueueProcessingState to use.
-     * @param dataFrameBuilder The DataFrameBuilder to use for constructing DataFrames.
-     */
-    private void processOperations(Queue<CompletableOperation> operations, QueueProcessingState state, DataFrameBuilder<Operation> dataFrameBuilder) {
-        try {
-            while (!operations.isEmpty()) {
-                CompletableOperation o = operations.poll();
-                if (processOperation(o, dataFrameBuilder)) {
-                    // Add the operation as 'pending', only if we were able to successfully append it to a data frame.
-                    // We only commit data frames when we attempt to start a new record (if it's full) or if we try to
-                    // close it, so we will not miss out on it.
-                    state.addPending(o);
+            } catch (Throwable ex) {
+                // Fail ALL the operations that haven't been checkpointed yet.
+                Throwable realCause = ExceptionHelpers.getRealException(ex);
+                this.state.fail(realCause, null);
+
+                if (isFatalException(realCause)) {
+                    // If we encountered a fatal exception, it means we detected something that we cannot possibly recover from.
+                    // We need to shutdown right away (this will be done by the main loop).
+
+                    // But first, fail any Operations that we did not have a chance to process yet.
+                    cancelIncompleteOperations(operations, realCause);
+                    throw realCause;
                 }
             }
-        } catch (Throwable ex) {
-            handleIterationException(ex, state, operations);
-            dataFrameBuilder.reset();
         }
     }
 
     /**
      * Processes a single operation.
      * Steps:
      * <ol>
-     * <li> Pre-processes operation (in MetadataUpdater)
-     * <li> Assigns Sequence Number
-     * <li> Appends to DataFrameBuilder
+     * <li> Pre-processes operation (in MetadataUpdater).
+     * <li> Assigns Sequence Number.
+     * <li> Appends to DataFrameBuilder.
      * <li> Accepts operation in MetadataUpdater.
      * </ol>
-     * Any exceptions along the way will result in the immediate failure of the operations. Exceptions do not bubble out
+     * Any exceptions along the way will result in the immediate failure of the operation. Exceptions do not bubble out
```

Comment:
The goal of this method is simple: Process the operation atomically (preprocess (metadata tx) + add to DataFrame + accept (into metadata txn)). If it fails to process (Metadata rejected it - which is almost always user error) the it returns false. If a more serious problem happened which was not due to this, an exception is thrown, which is handled in the parent method. 

I wanted to separate the user-errors from the actual Tier1 processing errors. I don't want to overload the user errors with yet another exception which wraps an inner exception and then handle that upstream. IMO, that would make the code look much uglier.